warehouse/query-core/src/main/java/datawave/query/util/sortedset/BufferedFileBackedSortedSet.java

```diff
@@ -156,56 +158,78 @@ public boolean containsAll(Collection<?> c) {
+            
+            for (int r = 0; r < rounds; r++) {
                 
-                MultiSetBackedSortedSet<E> newSet = new MultiSetBackedSortedSet<>();
+                // create a set for those sets to be compacted into one file
                 MultiSetBackedSortedSet<E> setToCompact = new MultiSetBackedSortedSet<>();
-                
-                int iteration = 0;
-                byte subtraction = (byte) (iteration < iterationsWithExtra ? 0 : 1);
-                for (int i = 0; i < numSets; i++) {
+                for (int i = 0; i < setsPerCompaction; i++) {
                     setToCompact.addSet(sets.get(i));
-                    if (setToCompact.getSets().size() == (setsPerCompaction - subtraction)) {
-                        newSet.addSet(compact(setToCompact));
-                        setToCompact.clear();
-                        iteration++;
-                        subtraction = (byte) (iteration < iterationsWithExtra ? 0 : 1);
-                    }
                 }
-                if (!setToCompact.isEmpty()) {
-                    if (log.isDebugEnabled()) {
-                        log.debug("Starting compaction for " + setToCompact);
-                    }
-                    long start = System.currentTimeMillis();
-                    FileSortedSet<E> compaction = compact(setToCompact);
-                    if (log.isDebugEnabled()) {
-                        long delta = System.currentTimeMillis() - start;
-                        log.debug("Compacted " + setToCompact + " -> " + compaction + " in " + delta + "ms");
-                    }
-                    newSet.addSet(compaction);
-                    setToCompact.clear();
+                
+                // remove the sets being compacted from our list
+                // creating new array list to allow previous list to be GC'ed
+                sets = new ArrayList<>(sets.subList(setsPerCompaction, sets.size()));
+                
+                // compact it
+                if (log.isDebugEnabled()) {
+                    log.debug("Starting compaction for " + setToCompact);
                 }
-                this.set = newSet;
+                long start = System.currentTimeMillis();
+                FileSortedSet<E> compaction = compact(setToCompact);
+                if (log.isDebugEnabled()) {
+                    long delta = System.currentTimeMillis() - start;
+                    log.debug("Compacted " + setToCompact + " -> " + compaction + " in " + delta + "ms");
+                }
+                
+                // add the compacted set to our final multiset
+                newSet.addSet(compaction);
+                
+                // clear the compactions set to remove the files that were compacted
+                setToCompact.clear();
+            }
+            
+            // now add in the sets we did not compact
+            for (int i = 0; i < sets.size(); i++) {
```

Comment:
Why not?
```java
newSet.addAll(sets);
```

Reply:
because addSet() is different than add()....